src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs

```diff
@@ -572,53 +574,67 @@ private record struct ExplicitMetadataLookup(long EntityId, string MetadataKey);
     }
 
     public async Task<GatewayModel.StateNonFungibleIdsResponse> NonFungibleIds(
-        IEntityStateQuerier.PageRequest request,
+        EntityAddress nonFungibleResourceAddress,
         GatewayModel.LedgerState ledgerState,
+        GatewayModel.ComplexKeysCursor? cursor,
+        int pageSize,
         CancellationToken token = default)
     {
-        var entity = await GetEntity<GlobalNonFungibleResourceEntity>(request.Address, ledgerState, token);
+        var entity = await GetEntity<GlobalNonFungibleResourceEntity>(nonFungibleResourceAddress, ledgerState, token);
 
         var cd = new CommandDefinition(
             commandText: @"
-WITH most_recent_non_fungible_id_store_history_slice (non_fungible_id_data_ids, non_fungible_ids_total_count) AS (
-    SELECT non_fungible_id_data_ids[@startIndex:@endIndex], cardinality(non_fungible_id_data_ids)
-    FROM non_fungible_id_store_history
-    WHERE from_state_version <= @stateVersion AND non_fungible_resource_entity_id = @entityId
+SELECT
+    d.id AS Id,
+    d.from_state_version AS FromStateVersion,
+    d.non_fungible_id AS NonFungibleId
+FROM non_fungible_id_definition d
+INNER JOIN LATERAL (
+    SELECT *
+    FROM non_fungible_id_history
+    WHERE non_fungible_id_definition_id = d.id AND from_state_version <= @stateVersion
     ORDER BY from_state_version DESC
     LIMIT 1
-)
-SELECT nfid.non_fungible_id AS NonFungibleId, hs.non_fungible_ids_total_count AS NonFungibleIdsTotalCount
-FROM most_recent_non_fungible_id_store_history_slice hs
-INNER JOIN non_fungible_id_data nfid ON nfid.id = ANY(hs.non_fungible_id_data_ids)
-ORDER BY array_position(hs.non_fungible_id_data_ids, nfid.id);
-",
+    ) h ON TRUE
+WHERE
+    d.non_fungible_resource_entity_id = @nonFungibleResourceEntityId
+  AND (d.from_state_version, d.id) <= (@cursorStateVersion, @cursorId)
+  AND d.from_state_version <= @stateVersion
+  AND h.is_deleted = false
+ORDER BY d.from_state_version DESC, d.id DESC
+LIMIT @limit
+;",
             parameters: new
             {
+                nonFungibleResourceEntityId = entity.Id,
                 stateVersion = ledgerState.StateVersion,
-                entityId = entity.Id,
-                startIndex = request.Offset + 1,
-                endIndex = request.Offset + request.Limit,
+                cursorStateVersion = cursor?.StateVersionBoundary ?? long.MaxValue,
```

Comment:
Correct me if I'm wrong but if I'm not mistaken any request with `cursor` must explicitly define `at_ledger_state.state_version`, isn't it? Then isn't the `state_version` inside the cursor object redundant? The `Id` on its own is enough to serve as a cursor.

Reply:
Well I think it should be an exception to have `cursor?.StateVersionBoundary` > `at_ledger_state.state_version`, in which case, we should use `cursor?.StateVersionBoundary ?? at_ledger_state.state_version` here. 

But I disagree that the `state_version` in the cursor is redundant - conceptually it's a composite cursor `(state_version_entry_added, entry_id)`.